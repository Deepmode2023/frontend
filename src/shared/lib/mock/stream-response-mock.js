const stateSessionMockSuccessfully = {
  ME: [{}],
  REFUSE: [{}],
  EXTENDING: [{}],
};

const stateSessionMockUnSuccessfully = {
  ME: [{}],
  REFUSE: [{}],
  EXTENDING: [{}],
};
