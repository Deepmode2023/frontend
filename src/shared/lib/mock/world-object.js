export default {
  Africa: {
    Egypt: {
      Cairo: [
        {
          street: "Tahrir Square",
          postalCode: "11599",
          latitude: 30.0444,
          longitude: 31.2357,
        },
        {
          street: "Al Haram",
          postalCode: "12556",
          latitude: 29.9773,
          longitude: 31.1325,
        },
      ],
      Alexandria: [
        {
          street: "Corniche Road",
          postalCode: "21617",
          latitude: 31.2001,
          longitude: 29.9187,
        },
        {
          street: "Abu Qir Street",
          postalCode: "21611",
          latitude: 31.2622,
          longitude: 30.0054,
        },
      ],
    },
    Nigeria: {
      Lagos: [
        {
          street: "Broad Street",
          postalCode: "101231",
          latitude: 6.4516,
          longitude: 3.3881,
        },
        {
          street: "Awolowo Road",
          postalCode: "101233",
          latitude: 6.4541,
          longitude: 3.4019,
        },
      ],
      Abuja: [
        {
          street: "Aso Drive",
          postalCode: "900271",
          latitude: 9.0765,
          longitude: 7.3986,
        },
        {
          street: "Garki Road",
          postalCode: "900242",
          latitude: 9.0579,
          longitude: 7.4896,
        },
      ],
    },
  },
  Europe: {
    France: {
      Paris: [
        {
          street: "Champs-Élysées",
          postalCode: "75008",
          latitude: 48.8566,
          longitude: 2.3522,
        },
        {
          street: "Rue de Rivoli",
          postalCode: "75004",
          latitude: 48.8556,
          longitude: 2.3609,
        },
      ],
      Lyon: [
        {
          street: "Rue de la République",
          postalCode: "69002",
          latitude: 45.7607,
          longitude: 4.8357,
        },
        {
          street: "Avenue des Frères Lumière",
          postalCode: "69008",
          latitude: 45.75,
          longitude: 4.8793,
        },
      ],
    },
    Germany: {
      Berlin: [
        {
          street: "Unter den Linden",
          postalCode: "10117",
          latitude: 52.517,
          longitude: 13.3889,
        },
        {
          street: "Kurfürstendamm",
          postalCode: "10719",
          latitude: 52.5026,
          longitude: 13.3325,
        },
      ],
      Munich: [
        {
          street: "Marienplatz",
          postalCode: "80331",
          latitude: 48.1374,
          longitude: 11.5755,
        },
        {
          street: "Leopoldstraße",
          postalCode: "80802",
          latitude: 48.1596,
          longitude: 11.5857,
        },
      ],
    },
  },
  Asia: {
    Japan: {
      Tokyo: [
        {
          street: "Shibuya Crossing",
          postalCode: "150-0002",
          latitude: 35.6595,
          longitude: 139.7006,
        },
        {
          street: "Ginza",
          postalCode: "104-0061",
          latitude: 35.6717,
          longitude: 139.765,
        },
      ],
      Osaka: [
        {
          street: "Dotonbori",
          postalCode: "542-0071",
          latitude: 34.6687,
          longitude: 135.5019,
        },
        {
          street: "Shinsaibashi",
          postalCode: "542-0085",
          latitude: 34.6762,
          longitude: 135.5011,
        },
      ],
    },
    China: {
      Beijing: [
        {
          street: "Wangfujing",
          postalCode: "100006",
          latitude: 39.9087,
          longitude: 116.3975,
        },
        {
          street: "Chang'an Avenue",
          postalCode: "100031",
          latitude: 39.9042,
          longitude: 116.4074,
        },
      ],
      Shanghai: [
        {
          street: "Nanjing Road",
          postalCode: "200001",
          latitude: 31.2304,
          longitude: 121.4737,
        },
        {
          street: "The Bund",
          postalCode: "200002",
          latitude: 31.24,
          longitude: 121.4905,
        },
      ],
    },
  },
};
