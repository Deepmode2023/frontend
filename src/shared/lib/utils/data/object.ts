import Stack from "../data-structure/stack/stack";
import get from "lodash.get";

type ObjectKeyType = string | number;

interface IGetPathInObjectProps {
  searchKey: ObjectKeyType;
  searchedObject: Record<ObjectKeyType, any>;
  withKey?: boolean;
  isAllMatches?: boolean;
}

interface IRecursiveObjectProps {
  obj: Record<ObjectKeyType, any>;
  searchKey: ObjectKeyType;
  path: Array<string>;
  pathInstance: Array<string>;
}

const returnPathWithKey = (
  withKey: boolean,
  path: string,
  searchedKey: IGetPathInObjectProps["searchKey"]
) => {
  const keyAsString = String(searchedKey);

  if (withKey && !path.includes(keyAsString)) {
    return `${path}.${keyAsString}`;
  }

  return withKey ? path : path.replace(`.${keyAsString}`, "");
};

const returnedPathToString = (
  prevPath: string,
  currentPath: string
): string => {
  const formattedPath = /^\d+$/.test(currentPath)
    ? `[${currentPath}]`
    : currentPath;
  return `${prevPath}.${formattedPath}`;
};

const getPathInObject = ({
  searchKey,
  searchedObject,
  withKey = false,
  isAllMatches = false,
}: IGetPathInObjectProps): Array<string> => {
  const pathsStack = new Stack<string>();
  const resultPaths: Array<string> = [];

  if (
    typeof searchedObject !== "object" ||
    !searchedObject ||
    (typeof searchKey !== "string" && typeof searchKey !== "number")
  ) {
    return [];
  }

  Object.keys(searchedObject)
    .reverse()
    .forEach((key) => pathsStack.push(key));

  while (!pathsStack.isEmpty()) {
    const path = pathsStack.pop()!;
    const deeplyObj = get(searchedObject, path);

    if (!path.includes(String(searchKey))) {
      if (deeplyObj && typeof deeplyObj === "object") {
        Object.keys(deeplyObj)
          .reverse()
          .forEach((deepKey) => {
            pathsStack.push(returnedPathToString(path, deepKey));
          });
      }
      continue;
    }

    resultPaths.push(returnPathWithKey(withKey, path, searchKey));

    if (!isAllMatches) {
      break;
    }
  }

  return resultPaths;
};

const getPath = (
  props: Omit<IGetPathInObjectProps, "isAllMatches">
): string => {
  return getPathInObject({ ...props, isAllMatches: false })[0] ?? "";
};

const getPaths = (
  props: Omit<IGetPathInObjectProps, "isAllMatches">
): Array<string> => getPathInObject({ ...props, isAllMatches: true });

export { getPaths, getPath };
