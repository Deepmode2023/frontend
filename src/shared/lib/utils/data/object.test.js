import { describe, test, expect } from "bun:test";
import { getPath, getPaths } from "./object";
import mock from "../../mock/world-object";

describe("Total testing 'getPath', 'getPaths' ", () => {
  const result = "";

  test("... & with wierd searchedKey", () => {
    expect(
      getPath({
        searchKey: [],
        searchedObject: mock,
        withKey: false,
      })
    ).toBe(result);
    expect(
      getPath({ searchKey: undefined, searchedObject: mock, withKey: false })
    ).toBe(result);

    expect(
      getPath({ searchKey: null, searchedObject: mock, withKey: true })
    ).toBe(result);

    expect(
      getPath({ searchKey: NaN, searchedObject: mock, withKey: false })
    ).toBe(result);
  });

  test("... & with wierd searchedObject", () => {
    expect(
      getPath({ searchKey: "street", searchedObject: [], withKey: true })
    ).toBe(result);
    expect(
      getPath({
        searchKey: "street",
        searchedObject: "test",
        withKey: true,
      })
    ).toBe(result);
    expect(
      getPath({ searchKey: 122323, searchedObject: [], withKey: false })
    ).toBe(result);
  });
});

describe("Testing 'getPath' with diffecult mock object", () => {
  test("... & existing key in mock data and path with key", () => {
    const searchKey = "Cairo";
    const result = "Africa.Egypt.Cairo";
    expect(getPath({ searchKey, searchedObject: mock, withKey: true })).toBe(
      result
    );
  });

  test("... & existing the key inside deeply array with key", () => {
    const searchKey = "street";
    const result = "Africa.Egypt.Cairo.[0].street";

    expect(
      getPath({
        searchKey,
        searchedObject: mock,
        withKey: true,
      })
    ).toBe(result);

    expect(
      getPath({ searchKey, searchedObject: mock, withKey: false })
    ).toMatchSnapshot();
  });

  test("... & pass unexisting key inside deeply with key", () => {
    const searchKey = "streetsss";
    const result = "";

    expect(
      getPath({
        searchKey,
        searchedObject: mock,
        withKey: true,
      })
    ).toBe(result);
  });

  test("... & pass unexisting key inside deeply without key", () => {
    const searchKey = "streetsss";
    const result = "";

    expect(
      getPath({
        searchKey,
        searchedObject: mock,
        withKey: false,
      })
    ).toBe(result);
  });

  test("... & pass without key", () => {
    const searchKey = "street";
    const result = "Africa.Egypt.Cairo.[0]";

    expect(
      getPath({
        searchKey,
        searchedObject: mock,
        withKey: false,
      })
    ).toBe(result);

    expect(
      getPath({ searchKey, searchedObject: mock, withKey: false })
    ).toMatchSnapshot();
  });
});

describe("Testing 'getPaths' with diffecult mock object", () => {
  const slicedMock = mock.Africa.Egypt;
  test("... & existing key in mock data and path with key", () => {
    const searchKey = "street";
    const result = [
      "Cairo.[0].street",
      "Cairo.[1].street",
      "Alexandria.[0].street",
      "Alexandria.[1].street",
    ];
    expect(
      getPaths({ searchKey, searchedObject: slicedMock, withKey: true })
    ).toEqual(result);

    expect(
      getPaths({ searchKey, searchedObject: slicedMock, withKey: true })
    ).toMatchSnapshot();
  });

  test("... & existing key in mock data and path without key", () => {
    const searchKey = "street";
    const result = [
      "Cairo.[0]",
      "Cairo.[1]",
      "Alexandria.[0]",
      "Alexandria.[1]",
    ];
    expect(
      getPaths({ searchKey, searchedObject: slicedMock, withKey: false })
    ).toEqual(result);

    expect(
      getPaths({ searchKey, searchedObject: slicedMock, withKey: false })
    ).toMatchSnapshot();
  });

  test("... & pass unexisting key inside deeply with key", () => {
    const searchKey = "streetsss";
    const result = "";

    expect(
      getPath({
        searchKey,
        searchedObject: slicedMock,
        withKey: true,
      })
    ).toBe(result);
  });
});
