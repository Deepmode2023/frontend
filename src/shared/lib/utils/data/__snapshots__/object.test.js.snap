// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing 'getPath' with diffecult mock object ... & existing the key inside deeply array with key 1`] = `"Africa.Egypt.Cairo.[0]"`;

exports[`Testing 'getPath' with diffecult mock object ... & pass without key 1`] = `"Africa.Egypt.Cairo.[0]"`;

exports[`Testing 'getPaths' with diffecult mock object ... & existing key in mock data and path with key 1`] = `
[
  "Cairo.[0].street",
  "Cairo.[1].street",
  "Alexandria.[0].street",
  "Alexandria.[1].street",
]
`;

exports[`Testing 'getPaths' with diffecult mock object ... & existing key in mock data and path without key 1`] = `
[
  "Cairo.[0]",
  "Cairo.[1]",
  "Alexandria.[0]",
  "Alexandria.[1]",
]
`;
